netcat: explicit use of IPv4 and IPv6

jensdrenhaus opened the pull request:
Fixes #3213

jensdrenhaus commented:
Still need to do final testing and improve some logging.

rminnich commented on cmds/core/netcat/connect.go:
```diff
@@ -28,6 +28,24 @@ import (
 func (c *cmd) connectMode(output io.Writer, network, address string) error {
+	if network == "tcp" || network == "udp" {
+		network4 := network + "4"
+		err4 := c.connect(output, network4, address)
```
or you can just
err4 := c.connect(output, network+"4", address)

rminnich commented on cmds/core/netcat/listen.go:
```diff
@@ -24,6 +24,24 @@ import (
+		if err6 == nil {
+			return nil
+		}
+		return fmt.Errorf("listen IPv4: %w\nlisten IPv6: %wgi", err4, err6)
```
I don't think you should embed newlines in an error message. You could use a single error variable and errors.Join I think.

rminnich replied:
also, does the err have information such that you don't need the "listen IPV4"?
I suspect so.
You might just need fmt.Errorf("listenMode:%w", and you could have a single err and  errors.Join

jensdrenhaus replied:
Nice, errors.Join seem decent!

jensdrenhaus commented:
BTW, any idea why we do not allow squash and merge? So that I could add commits addressing comments and such without the need to force push and thus messing up the PR conversations history.

leongross commented on cmds/core/netcat/connect_test.go:
```diff
@@ -20,6 +22,113 @@ import (
+		network     string
+		address     string
+		connectFunc func(output io.Writer, network, address string) error
+		expectError bool
```
If we have test cases that throw known errors, wouldn't it make more sense to embed them into the test cases? So either it is `nil` or the error we predefined?

 ```suggestion
	     expectedError error
```

and then we can check for the result
```go
if err := cmd.connectMode(io.Discard, tt.network, tt.address); err != tt.expectedError {
    t.Errorf("wanted %v, got %v", tt.expectedError, err)
}
```

jensdrenhaus replied:
If there are defined error values, yes, sure. This would make sense in the actual [connect func](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-1445756be1283b81274f42febc4561c59c56b8d236f6b49317e23eb5d394c3edR46) which is the former 'ConnectMode()' function. 

What I did here is pushing the code of the former 'ConnectMode()' function into 'connect()' and added logic in the new 'ConnectMode()' function to call 'connect()' several times in some cases. To test this logic I'm using a stub function instead of 'connect()' that just return errors as needed. I think this is sufficient to test the introduced logic. 

Adding defined error types and testing against them would be in the scope of [TestConnect](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-16aa8d474e6f1d06d5edc08b989c613ff40f2854257292332a816e3f325a8a9eR152) and is out of scope for this PR.

leongross replied:
I understand. Let's keep it the way it is.

leongross commented on cmds/core/netcat/listen_test.go:
```diff
@@ -18,6 +18,113 @@ import (
+		network     string
+		address     string
+		listenFunc  func(output io.Writer, network, address string) error
+		expectError bool
```
Same here
```suggestion
		expectedError error
```

jensdrenhaus replied:
See comment on [TestConnectMode](https://github.com/u-root/u-root/pull/3291/files/2c13f0513578ad8168ab7f2c0a4ba427f78c643f#diff-16aa8d474e6f1d06d5edc08b989c613ff40f2854257292332a816e3f325a8a9eR25)

jensdrenhaus commented:
@rminnich ready for review. Your change request should be addressed now.

rminnich commented on cmds/core/netcat/connect_test.go:
```diff
@@ -20,6 +22,110 @@ import (
+		network       string
+		address       string
+		connectFunc   func(output io.Writer, network, address string) error
+		expectedError error
```
one little note here: go style tends to be brief, it's more comment to name this err and not expectedError.
See what you think. I personally prefer just err.

rminnich commented on cmds/core/netcat/connect_test.go:
```diff
@@ -20,6 +22,110 @@ import (
+			}
+			err := cmd.connectMode(io.Discard, tt.network, tt.address)
+			if !errors.Is(err, tt.expectedError) {
+				t.Errorf("Expected error %v, got %v", tt.expectedError, err)
```
The common pattern is the 'got, want' pattern, I believe, i.e. what you got, and what you expected. 
Certainly in my code reviews at Google that's what they'd tell me. 

It makes it easier to scan the logs in Ci if we keep this consistent. again, I leave it to you.

jensdrenhaus replied:
Agree, consistency matters.

rminnich commented on cmds/core/netcat/netcat.go:
```diff
@@ -335,6 +335,10 @@ type cmd struct {
 	stderr io.Writer
 	config *netcat.Config
 	args   []string
+
```
This is now how it should be done, sorry. 

The existence of these indicated a problem with the design. Let's reconsider this.

jensdrenhaus replied:
Yeah, I was wondering if this was a wise approach or not. Any alternative suggestions would be appreciated. 

The thing is that the _listen_ and _connect_ functions of netcat are basically wrappers around some net package functionality (or tls package, et al, that work with `net.Conn` and `net.Listener`). I think we should not test the functionality of the net package here, but rather isolate the logic added by netcat and test that remaining part. 

I have not delved into refactoring the existing connect and listen functions and their testing. Pretty sure there is potential for improvement in the above manner. 
To me, the way I added the logic of the change seems somewhat decent, and I was looking for a way to add meaningful testing for that added logic. 

Again, glad to see an alternative approach.

rminnich reviewed (changes requested):
connectStubFn and listenStubFn are not a good way to do this, let's slow down a bit and get it right.

rminnich commented:
create a mockCmd struct that wraps command
implement connect for that struct
something like
struct mockCmd{
cmd
network string
}

func(m*mockCmd) connect(...)error{
if m.network == network
return nil
return connectErr
}
or some such

do the same for the other function

see mocks at end of the test file as an example

jensdrenhaus commented:
> create a mockCmd struct that wraps command implement connect for that struct something like struct mockCmd{ cmd network string }
> 
> func(m*mockCmd) connect(...)error{ if m.network == network return nil return connectErr } or some such
> 
> do the same for the other function
> 
> see mocks at end of the test file as an example

Thanks, @rminnich for that input, always fascinated by the power of embedding in Go. 
Unfortunately it does not work out here. 

I want to test cmd.connectMode() function. This function is calling cmd.connect(), so I need to mock/inject cmd.connect(). 

If I use your approach like so: 

```
type mockCmd struct {
	cmd
	connectStubFn func(output io.Writer, network, address string) error
}

func (m *mockCmd) connect(output io.Writer, network, address string) error {
	if m.connectStubFn == nil {
		return errors.New("mockCmd: connectStubFn is not set")
	}
	return m.connectStubFn(output, network, address)
}

func TestConnectMode(t *testing.T) {
	connectErr := errors.New("failed to connect")

	tests := []struct {
		name        string
		network     string
		address     string
		connectFunc func(output io.Writer, network, address string) error
		err         error
	}{
		{
			name:    "TCPv4 success",
			network: "tcp",
			address: "localhost:8080",
			connectFunc: func(output io.Writer, network, address string) error {
				if network == "tcp4" {
					return nil
				}
				return connectErr
			},
			err: nil,
		},

// ...

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			cmd := &mockCmd{
				connectStubFn: tt.connectFunc,
			}
			err := cmd.connectMode(io.Discard, tt.network, tt.address)
			if !errors.Is(err, tt.err) {
				t.Errorf("got %v, want %v", err, tt.err)
			}
		})
	}

}
```
mockCmd.connect will not be called but cmd.connect(). 

Pls correct me if I misunderstood. 

What about classic DI like: 
```
type connectFn func(io.Writer, string, string) error

func (c *cmd) connectMode(output io.Writer, network, address string, fn connectFn) error {
...
}
```

MDr164 commented:
> BTW, any idea why we do not allow squash and merge? So that I could add commits addressing comments and such without the need to force push and thus messing up the PR conversations history.

It messes up the merge commit message, at least that was the reason I remember was the most crucial one. But GitHub allows you to see differences quite easily even after a force push to the branch imo.
